### ISSUE TITLE
Security: Freeze both our and the dev's package with package-lock.json

### ISSUE THREAD

**Issue by plocket** [plocket/docassemble-cucumber#614]
I thought we had this issue, but I can't find it. This will ensure fullest security so that the updating/lack of security of upstream packages doesn't affect our packages except when we deliberately update package-lock.json.

For our dev's repo we need to:
- create a file to store their package-lock.json
- use/copy that file to create the dev's package-lock.json in the action (instead of the package.json)

For us and for the dev: configure our repo and their repo (through our action) to install using the package-lock.json.

**Comment by plocket** [comment]
For copying a file from our own repository, possibly see `action_path` at https://docs.github.com/en/actions/learn-github-actions/contexts.

**Comment by plocket** [comment]
This little test in our action works in a dependent repo:
```yml
runs:
  using: "composite"
  steps:
    - name: make file
      run: |
        touch cucumber.js
        cat "${{ github.action_path }}/cucumber.js" > cucumber.js
        file=$(<"cucumber.js")
        echo $file
      shell: bash
```

Note: tmate can't access values of variables like `github.whatever` in your local terminal. E.g. `github.action_path`. It gets the error 'bad substitution'. Instead, you have to (in your action.yml) add an environment variable to contain that value and use it later. Something like:

```yml
    - run: echo "ACTION_PATH=${{ github.action_path }}" >> $GITHUB_ENV
      shell: bash
    - name: make file
      run: |
        touch cucumber.js
        cat "${ACTION_PATH}/cucumber.js" > cucumber.js
        file=$(<"cucumber.js")
        echo $file
      shell: bash
```

**Comment by plocket** [comment]
How to use package-lock.json instead of package.json when doing `npm install`: https://blog.npmjs.org/post/171556855892/introducing-npm-ci-for-faster-more-reliable

**Comment by plocket** [comment]
We can store the package-lock.json in its own folder internally. When we want to update, we make changes in that internal file and run the usual `npm install` from that internal folder to update the package-lock.json (I believe).

**Comment by plocket** [comment]
Not sure how we didn't see this before - I'm not sure how/if we can freeze the developer's package-lock.json while still giving them the option of which npm version of ALKiln to install.

We currently allow the developer to choose their version of ALKiln to install with npm using their ALKILN_VERSION env var. This lets them use npm version `^4.0.0` or `4.3.1` and such:

```yml
      run: |
        cat << EOF > "package.json"
        {
          # scripts...
          "dependencies": {
            "colors": "1.4.0",
            "@cucumber/cucumber": "7.3.2",
            "@suffolklitlab/alkiln": "$ALKILN_VERSION"
          }
        }
        EOF
```

So if they're changing the version of ALKiln to install, they need to install from scratch every time.

Problematic possible solution: Maybe we can control it all through the action version? We'd hard-code both the package.json (with its current ALKiln version) and the package-lock.json. When they do `uses: SuffolkLITLab/ALKiln@fa962fb38a9d56b0903f36dbc6a310c573466538` they'd be picking their npm version that way.

The problem with that: We'd then end up needing to make a package-lock.json for every version of ALKiln and doesn't that partly defeat the purpose of what we're doing? It'd add a little security, but a lot less than we were hoping. The developer's package-lock.json wouldn't always change, but it would be changed with every change of ALKiln's version. Is that good enough?

The only other problem with it, I think, is that it would be even more of a pain to release each version. If we do this, we might want to push up the deadline on that "npm publish" workflow we're going to make... [Maybe as part of this, actually. The npm publish workflow issue  is #105 ]

[Also not sure how to keep track of which commits save which versions so we can tell developers which commit to refer to in their workflow (if they're trying to freeze versions). Maybe this really is a place for tags? Does that make sense?]

That's the only idea I've come up with so far.

Would love your thoughts on this, @BryceStevenWilley

**Comment by BryceStevenWilley** [comment]
In general, I agree that this makes releases more difficult for us, because we'd be taking on the burden of vetting all the dependencies (or at least testing them first). We can't get around that part I'm afraid; if we want the dependencies to be frozen and vetted before upgrading, then it needs to be either us or users vetting them.

> Maybe this really is a place for tags?

That's the perfect place for tags IMO, passing around commit hashes is definitely subpar, but `SuffolkLITLab/ALKiln@v4.3.2` is much better.

> We'd then end up needing to make a package-lock.json for every version of ALKiln and doesn't that partly defeat the purpose of what we're doing?

Maybe I'm confused, but I thought that's exactly what we're trying to do. We're freezing dependencies because we think that it's more likely that dependencies will have malicious updates than will have security fixes (or at least that we are more afraid of the consequences of malicious updates than we excited for security fixes, rightfully so). We aren't the only players in the `npm` space, and it's highly likely that others will be finding malicious releases before we do (which is what happened for `colors`). Using a `package-lock.json` greatly lowers the chance that we happen to update to a malicious package before others find and report it (and when we hear about it).

I think this is the right solution though: using only action versions (which implictly use the correct ALKiln version), and no longer using the `ALKILN_VERSION` variable.

**Comment by plocket** [comment]
Good points, I just hadn't thought all the way through the number of times we'd be updating various package-lock.json files. I thought it would be less often. Maybe that's just for our repo's dependencies, though, not for the action.

> we'd be taking on the burden of vetting all the dependencies (or at least testing them first)
...
> greatly lowers the chance that we happen to update to a malicious package before others find and report it (and when we hear about it)

I thought the goal was the second one, not the first. If you do envision vetting for each new version, what are your thoughts on the process for that?

> passing around commit hashes is definitely subpar, but SuffolkLITLab/ALKiln@v4.3.2 is much better.

People would still need to use the commit if they really want to freeze things as the contents of tags can change. It would give a marker, though, of which commit to use. Or am I misunderstanding something about tags?

**Comment by BryceStevenWilley** [comment]
> I thought the goal was the second one, not the first

Apologies, you're right. I'm using "vetting" too loosely, I do think that we'd test and check if dependencies work when updating though, which is what I had been trying to express. That means we won't update to completely breaking packages, but there's still a change that some package is doing something silly like crypto mining in the background.

> People would still need to use the commit if they really want to freeze things as the contents of tags can change

Sure. I did look it up and realized that Github has [protected tags](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-tag-protection-rules), which would only let admins (i.e. us) remove tags. Given this feature, I think that most people should be fine with using tags. If orgs are paranoid that we are going to force-remove and replace an existing tag with a malicious release, then they'll know enough to use the commit hash.

**Comment by BryceStevenWilley** [comment]
Remembering that we've had this discussion before, and landed on `shrinkwrap` as the right solution: https://github.com/SuffolkLITLab/ALKiln/discussions/532.

**Comment by plocket** [comment]
Discussions continuing in #105. Shrinkwrap may not be the answer we're looking for.

**Comment by BryceStevenWilley** [comment]
Conversation we had about shrinkwrap:

In the [npm docs](https://docs.npmjs.com/cli/v8/configuring-npm/npm-shrinkwrap-json), it's stated that:

> The recommended use-case for npm-shrinkwrap.json is applications deployed through the publishing process on the registry: for example, daemons and command-line tools intended as global installs or devDependencies

I think that ALKiln very much fits the description of command-line tools, as we haven't been implementing Kiln as a library made for external use.

With shrinkwrap, we could switch from:
* creating a `package.json` or `package-lock.json` file, and then calling `npm install` to install all of our dependencies

to:
* calling `npm install -g` to globally install ALKiln, without the need to be in an npm packgae.

@plocket had previously run into issues with using global installs, where the `bin` files of dependencies aren't installed. We use `cucumber-js`, which is a binary in the cucumber package. I think this [stack overflow issue](https://stackoverflow.com/a/14657796/11416267) describes it. However, it is possible to not directly use `cucumber-js` directly by instead writing our own small JS wrapper over cucumber's CLI runner. See https://github.com/SuffolkLITLab/ALKiln/tree/lighter_package_json, and https://github.com/SuffolkLITLab/ALKiln/blob/lighter_package_json/lib/run_cucumber.js specifically.

Using `npm install -g` in this way would require a change to people's `action.yml`s, and the entire install process is being simplified. This also has implications for [our release workflow](https://github.com/SuffolkLITLab/ALKiln/issues/105), which will be detailed in that issue.

**Comment by plocket** [comment]
Just a note that our users do still need their own scripts, so we still may need them to have a package.json. [Maybe that still fits into what the comment above describes and I've misunderstood something.]

**Comment by BryceStevenWilley** [comment]
> Just a note that our users do still need their own scripts, so we still may need them to have a package.json

Good catch, I think I mentioned in our conversation but didn't record it here: to make this change we would need to turn most of the npm scripts into standalone `*.js` files in that we install into `/bin`. This makes it such that instead of running `npm run langs_setup`, we can just run `alkiln-langs-setup` instead, so we don't need the npm scripts. The scripts that we use for running internal tests (like `unit`) can and should remain, but anything that we run in the `action.yml` workflow should be a bin.

This also has the added benefit of addressing https://github.com/SuffolkLITLab/ALKiln/issues/594, since we use non-Windows specific bash functionality (making a function and passing around arguments with `$@`) in many of the scripts.

Those scripts are currently: 

```json
  "scripts": {
    "test": "npm run langs_setup && npm run cucumber -- $@",
    "cucumber_base": "./node_modules/.bin/cucumber-js --require ./node_modules/@suffolklitlab/alkiln/lib/index.js",
    "cucumber": "run_cucumber(){ npm run cucumber_base -- \"$@\" docassemble/*/data/sources/*.feature --retry 1; }; run_cucumber",
    "langs": "npm run langs_setup && ./node_modules/.bin/cucumber-js --require ./node_modules/.bin/alkiln-cucumber docassemble/*/data/sources/*_lang*.feature",
    "langs_setup": "node_modules/.bin/alkiln-langs 'docassemble/*/data/sources/*.feature'",
    "setup": "node_modules/.bin/alkiln-setup",
    "takedown": "node_modules/.bin/alkiln-takedown"
  },
 ```
 
 The only scripts that need to be replaced by JS files are `langs`/`langs_setup`, and `cucumber`. `cucumber_base` has been replaced in https://github.com/SuffolkLITLab/ALKiln/blob/lighter_package_json/lib/run_cucumber.js. The other scripts are combinations of the above scripts (test, and some other internal ones) and `setup` and `takedown` are already installed scripts that we can call directly as `alkiln-setup` when it's globally installed.
 
 An approach for this would be to turn the npm scripts into JS files, and if they are called without arguments, use the arguments passed above (i.e. `docassemble/*/data/sources/*.feature --retry 1`). If you call them with arguments, use the passed in arguments. That allows use to still have internal testing scripts use the same JS files, since many of our internal testing scripts are the same as these, with just different arguments.

**Comment by BryceStevenWilley** [comment]
This was merged in v5 through #697. Going to close.
